src/Controller.cpp

```diff
@@ -63,12 +66,33 @@ extern "C"
         return (void *) (long) geopm_run_imp((struct geopm_ctl_c *)args);
     }
 
-    int geopmctl_main(void)
+    int geopmctl_main(int argc, char **argv)
     {
         int err = 0;
+
         try {
-            geopm::Controller ctl;
-            err = geopm_ctl_run((struct geopm_ctl_c *)&ctl);
+#ifdef GEOPM_ENABLE_MPI
+            bool do_ctl_local = geopm::environment().do_ctl_local();
+            if (!do_ctl_local) {
+                err = PMPI_Init(&argc, &argv);
+                if (err) {
+                    int str_size = MPI_MAX_ERROR_STRING;
+                    char error_str[MPI_MAX_ERROR_STRING + 1] = {};
+                    PMPI_Error_string(err, error_str, &str_size);
+                    fprintf(stderr, "Error: %s\n", error_str);
+                    return err;
+                }
+            }
+#endif
+            {
+                std::unique_ptr<geopm::Controller> ctl = std::make_unique<geopm::Controller>();
+                ctl->run();
+            }
```

Comment:
Was the extra {} here necessary for something?

Reply:
that forces the controler destructor to be called, which needs to happen before MPI_FInalize()